src/com/xilinx/rapidwright/design/tools/NetlistClockDetection.java

```diff
@@ -0,0 +1,180 @@
+        seriesPrimsToClockPins.put(Series.Versal, versalPrimsToClockPins);
+    }
+
+    /**
+     * Finds all terminal hierarchical port insts (pins) which gate changes to the given
+     * hierarchical net's value, i.e., finds all terminal pins which may be drivers of
+     * the net through non-clocked primitives and all likely terminal clock signal pins
+     * that gate signals which drive the net through clocked primitives.
+     *
+     * For example, when run on the output hierarchical net of a LUT2 whose inputs come
+     * from a flop and a LUT, this will return the terminal hierarchical port inst(s)
+     * driving the flop's clock pin as well as any gating drivers of the input LUT's
+     * inputs, determined recursively.
+     *
+     * @param hNet             The hierarchical net to find gating drivers of.
+     * @param primsToClockPins A map from primitive names to their input clock pins, if
+     *                         the primitive does have input clock pins. This is
+     *                         specific to the FPGA series.
+     * @param encountered      A set of already-encountered EDIFHierNets to skip during
+     *                         netlist traversal.
+     */
+    static Set<EDIFHierPortInst> getGatingDrivers(EDIFHierNet hNet, Map<String, Set<String>> primsToClockPins, Set<EDIFHierNet> encountered) {
+        Set<EDIFHierPortInst> out = new HashSet<EDIFHierPortInst>();
+
+        for (EDIFHierPortInst hPI : hNet.getLeafHierPortInsts(true, false)) {
+            if (hPI.getPortInst().getCellInst() == null) {
+                out.add(hPI);
+            } else {
+                Set<String> clkPins = primsToClockPins.get(hPI.getPortInst().getCellInst().getCellType().getName());
+                EDIFHierNet sourceHierNet;
+                EDIFHierCellInst leafCellInst = hPI.getHierarchicalInst().getChild(hPI.getPortInst().getCellInst());
+
+                List<EDIFHierPortInst> leafCellInstInputs =
+                    leafCellInst.getHierPortInsts().stream().filter(EDIFHierPortInst::isInput).collect(Collectors.toList());
+
+                if (!leafCellInstInputs.isEmpty()) {
+                    for (EDIFHierPortInst leafCellInstInput : leafCellInstInputs) {
+                        if (clkPins == null || clkPins.contains(leafCellInstInput.getPortInst().getName())) {
+                            if ((sourceHierNet = leafCellInstInput.getHierarchicalNet()) != null && !encountered.contains(sourceHierNet)) {
+                                encountered.add(sourceHierNet);
+                                out.addAll(getGatingDrivers(sourceHierNet, primsToClockPins, encountered));
+                            }
+                        }
+                    }
+                } else {
+                    out.add(hPI);
+                }
+            }
+        }
+
+        return out;
+    }
+
+    static Set<EDIFHierPortInst> getGatingDrivers(EDIFHierNet hNet, Map<String, Set<String>> primsToClockPins) {
```

Comment:
```suggestion
    public static Set<EDIFHierPortInst> getGatingDrivers(EDIFHierNet hNet, Map<String, Set<String>> primsToClockPins) {
```
Seems like we should make this available publicly?